src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,422 @@
+    (ideal.homogeneous_core 𝒜 :
+      ideal A → homogeneous_ideal 𝒜) :=
+{ choice := λ I HI, ⟨I, begin
+    have eq : I = I.homogeneous_core' 𝒜,
+    refine le_antisymm HI _,
+    apply (ideal.homogeneous_core'_le_ideal 𝒜 I),
+    rw eq, apply ideal.is_homogeneous.homogeneous_core,
+  end⟩,
+  gc := ideal.homogeneous_core.gc 𝒜,
+  u_l_le := λ I, by apply ideal.homogeneous_core'_le_ideal,
+  choice_eq := λ I H, le_antisymm H (I.homogeneous_core'_le_ideal _) }
+
+lemma ideal.homogeneous_core_eq_Sup :
+  I.homogeneous_core' 𝒜 = Sup {J : ideal A | J.is_homogeneous 𝒜 ∧ J ≤ I} :=
+begin
+  refine (is_lub.Sup_eq _).symm,
+  apply is_greatest.is_lub,
+  have coe_mono : monotone (coe : {I : ideal A // I.is_homogeneous 𝒜} → ideal A) := λ _ _, id,
+  convert coe_mono.map_is_greatest (ideal.homogeneous_core.gc 𝒜).is_greatest_u using 1,
+  simp only [subtype.coe_image, exists_prop, mem_set_of_eq, subtype.coe_mk],
+end
+
+end homogeneous_core
+
+section homogeneous_hull
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A]
+variables [algebra R A] [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--For any `I : ideal R`, not necessarily homogeneous, `I.homogeneous_hull' 𝒜` is
+the smallest homogeneous ideal containing `I`.-/
+def ideal.homogeneous_hull : homogeneous_ideal 𝒜 :=
+⟨ideal.span {r : A | ∃ (i : ι) (x : I), (graded_algebra.decompose 𝒜 x i : A) = r}, begin
+  rw ideal.is_homogeneous.iff_exists,
+  use {x : homogeneous_submonoid 𝒜 | ∃ (i : ι) (r : I), (graded_algebra.decompose 𝒜 r i : A) = x},
+  congr,
+  ext r,
+  split;
+  intros h,
+  { obtain ⟨i, ⟨x, hx1⟩, hx2⟩ := h,
+    exact ⟨⟨_, is_homogeneous_coe _⟩, ⟨⟨i, ⟨⟨x, hx1⟩, rfl⟩⟩, hx2⟩⟩,},
+  { obtain ⟨_, ⟨⟨i, ⟨⟨r, hr⟩, h⟩⟩, rfl⟩⟩ := h,
+    use i, use ⟨r, hr⟩, exact h }
+end⟩
+
+lemma ideal.ideal_le_homogeneous_hull :
+  I ≤ ideal.homogeneous_hull 𝒜 I :=
+begin
+  intros r hr,
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw [←graded_algebra.sum_support_decompose 𝒜 r],
+  refine ideal.sum_mem _ _, intros j hj,
+  apply ideal.subset_span, use j, use ⟨r, hr⟩, refl,
+end
+
+lemma ideal.homogeneous_hull_is_mono : monotone (ideal.homogeneous_hull 𝒜) := λ I J I_le_J,
+begin
+  apply ideal.span_mono,
+  rintros r ⟨hr1, ⟨x, hx⟩, rfl⟩,
+  refine ⟨hr1, ⟨⟨x, I_le_J hx⟩, rfl⟩⟩,
+end
+
+lemma ideal.homogeneous_hull_eq_Inf :
+  ideal.homogeneous_hull 𝒜 I = Inf { J : homogeneous_ideal 𝒜 | I ≤ J } :=
+begin
+  ext,
+  split;
+  intros hx,
+  { erw ideal.mem_Inf,
+    rintros _ ⟨K, HK1, rfl⟩,
+    erw [ideal.mem_span] at hx,
+    apply hx K,
+    rintros r ⟨i, ⟨⟨y, hy⟩, rfl⟩⟩,
+    exact K.2 _ (HK1 hy), },
+  { erw ideal.mem_Inf at hx,
+    refine @hx (ideal.homogeneous_hull 𝒜 I) _,
+    exact ⟨ideal.homogeneous_hull _ _, ideal.ideal_le_homogeneous_hull _ _, rfl⟩, }
+end
+
+lemma homogeneous_hull_eq_supr :
+  (I.homogeneous_hull 𝒜).1 = ⨆ i, ideal.span (graded_algebra.proj 𝒜 i '' I) :=
```

Comment:
Is `(graded_algebra.proj 𝒜 i '' I)` also homogenous?

Reply:
It may not be an ideal.

Let `I` be `(x + y^2)`, then `(graded_algebra.proj 𝒜 2 '' I)` contains `y^2` and `x^2` (coming from `(x + y^2)^2`), but presumably not `x^2 + y^2`.

Reply:
Sorry, I meant to say `ideal.span (graded_algebra.proj 𝒜 i '' I)`.

Reply:
yes, should we prove
```lean
lemma homogeneous_hull_eq_supr :
  (I.homogeneous_hull 𝒜) =
  ⨆ i, ⟨ideal.span (graded_algebra.proj 𝒜 i '' I), ideal.is_homogeneous_span 𝒜 _ (begin
    rintros _ ⟨x, h, rfl⟩,
    refine ⟨i, submodule.coe_mem _⟩,
  end)⟩ :=
```
instead

Reply:
I have made two versions
```lean
lemma homogeneous_hull_eq_supr :
   ↑(I.homogeneous_hull 𝒜) = ⨆ i, ideal.span (graded_algebra.proj 𝒜 i '' I) := _

lemma homogeneous_hull_eq_supr' :
  (I.homogeneous_hull 𝒜) =
  ⨆ i, ⟨ideal.span (graded_algebra.proj 𝒜 i '' I), ideal.is_homogeneous_span 𝒜 _ (begin
    rintros _ ⟨x, h, rfl⟩,
    refine ⟨i, submodule.coe_mem _⟩,
  end)⟩ := _
```